Realm/RLMQueryUtil.mm

```diff
@@ -1684,23 +1683,29 @@ bool is_self_value_for_key_path_function_expression(NSExpression *expression)
 void QueryBuilder::apply_map_expression(RLMObjectSchema *objectSchema, NSExpression *functionExpression,
                                         NSComparisonPredicateOptions options, NSPredicateOperatorType operatorType,
                                         NSExpression *right) {
+    std::vector<PathElement> pathElements;
+    get_path_elements(pathElements, functionExpression);
+
     NSString *keyPath;
-    NSString *mapKey;
-    if (functionExpression.operand.expressionType == NSKeyPathExpressionType) {
-        NSExpression *mapItems = [functionExpression.arguments firstObject];
-        NSExpression *linkCol = [[functionExpression.operand arguments] firstObject];
-        NSExpression *mapCol = [mapItems.arguments firstObject];
-        mapKey = [mapItems.arguments[1] constantValue];
-        keyPath = [NSString stringWithFormat:@"%@.%@", linkCol.keyPath, mapCol.keyPath];
-    } else {
-        keyPath = [functionExpression.arguments.firstObject keyPath];
-        mapKey = [functionExpression.arguments[1] constantValue];
-    }
+    NSExpression *keyPathExpression = functionExpression;
 
+    // We consider only `NSKeyPathExpressionType` values to build the keypath.
+    for (unsigned long i = 0; i < pathElements.size(); i++) {
+        if (keyPathExpression.arguments[0].expressionType == NSKeyPathExpressionType) {
+            keyPath = [NSString stringWithFormat:@"%@", keyPathExpression.arguments[0]];
```

Comment:
This appears to be a very roundabout way of writing `keyPathExpression.arguments[0].description`. I think the actually correct thing here is `keyPathExpression.arguments[0].keyPath`